[receiver/netflow] Netflow receiver implementation - PR 2

dlopes7 opened the pull request:
#34164 added the skeleton for the netflow receiver, this adds the implementation along with tests

* Implement the receiver construction, along with the UDP listener that receives the packets
* Implement the producer, the function that is responsible for parsing the data and sending it to the consumers
* Add tests

**Link to tracking Issue:** #32732

dlopes7 committed:
#32732 - Initial implementation

This adds the implementation for netflow receiver
Introduces:

* The OtelLogsProducerWrapper
* A parser function to convert proto to otel semantics
* The listener implementation as well as the error handling function

dlopes7 committed:
#32732 - Make internal methods private

dlopes7 committed:
#32732 - netflow - make checks

dlopes7 committed:
#32732 - netflow - simplify implementation, remove listener

dlopes7 committed:
netflow - add parser tests

dlopes7 committed:
netflow - add more tests

dlopes7 committed:
netflow - add test to check for udp receiver creation

dlopes7 committed:
netflow - gofmt

dlopes7 committed:
Merge remote-tracking branch 'upstream/main' into netflow-receiver-implementation

dlopes7 committed:
netflow - make checks

dlopes7 committed:
netflowreceiver - add changelog for implementation

dlopes7 committed:
netflow - fix license header

dlopes7 committed:
netflowreceiver - lint

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

dlopes7 committed:
Merge remote-tracking branch 'upstream/main' into netflow-receiver-implementation

evan-bradley commented on .chloggen/netflow-receiver-implementation.yaml:
```diff
@@ -0,0 +1,27 @@
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: Implement the UDP listener and the producer, as well as adding tests.
```
```suggestion
subtext: The receiver now supports receiving NetFlow v5, NetFow v9, IPFIX, and sFlow v5 logs.
```
Feel free to word this how you like, but I think we should just cover the user-facing changes here.

evan-bradley commented on receiver/netflowreceiver/receiver.go:
```diff
@@ -5,25 +5,165 @@ package netflowreceiver // import "github.com/open-telemetry/opentelemetry-colle
+
+	var decodeFunc utils.DecoderFunc
+	var p utils.FlowPipe
+	switch nr.config.Scheme {
```
Could we explain in the readme what protocol support users can expect when configuring one of these three schemes? I had to dig into the GoFlow2 source to understand what is happening here.

dlopes7 committed:
netflow - update .chloggen/netflow-receiver-implementation.yaml

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
netflow - remove scaffold issue

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
netflow - add comment about default queue size value

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
netflow - fix duplicate test case

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
netflow - fix typo

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

evan-bradley commented on receiver/netflowreceiver/parser.go:
```diff
@@ -0,0 +1,139 @@
+
+	// Construct the actual log record based on the otel semantic conventions
+	// see https://opentelemetry.io/docs/specs/semconv/general/attributes/
+	otelMessage := OtelNetworkMessage{
```
I dug into it some more and I see that the network semconv isn't necessarily a 1:1 fit here. I think we should do as much as we can, and we can create attributes for the pieces that don't fit so long as they're documented.

dlopes7 replied:
Implemented [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36865/commits/053543c7dffedf61df021a0e4b65dceafbca9a09)

dlopes7 committed:
netflow - better error handling messages

dlopes7 committed:
netflow - use a simple log record instead of struct

Use the semconv constants for the attribute keys where applicable

dlopes7 committed:
netflow - use context.Background() on producer

dlopes7 committed:
netflow - type the dropHandler interface

dlopes7 committed:
Update .chloggen/netflow-receiver-implementation.yaml

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
Merge branch 'netflow-receiver-implementation' of github.com:dlopes7/opentelemetry-collector-contrib into netflow-receiver-implementation

dlopes7 committed:
netflow - more concise startup log message

dlopes7 commented on receiver/netflowreceiver/receiver.go:
```diff
@@ -5,25 +5,165 @@ package netflowreceiver // import "github.com/open-telemetry/opentelemetry-colle
+
+		case errors.Is(err, debug.PanicError):
+			var pErrMsg *debug.PanicErrorMessage
+			if errors.As(err, &pErrMsg) {
```
Implemented in [this commit](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36865/commits/d68e82628f7072b980170637da64f98fecb2f097)

I found hard to not use the word panic because of the wrapper and types we are using from GoFlow2

evan-bradley replied:
What's here looks good to me. It's ultimately understandable if we mention panicking somewhere since that's what happens in the library, but ideally we should call out that it's an addressed error and move on.

dlopes7 commented on receiver/netflowreceiver/receiver.go:
```diff
@@ -5,25 +5,165 @@ package netflowreceiver // import "github.com/open-telemetry/opentelemetry-colle
+				nr.logger.Error("receiver stacktrace", zap.String("stack", string(pErrMsg.Stacktrace)))
+				nr.logger.Error("receiver msg", zap.Any("error", pErrMsg.Msg))
+			}
+			nr.logger.Error("receiver panic", zap.Error(err))
```
Implemented in [this commit](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36865/commits/d68e82628f7072b980170637da64f98fecb2f097)

dlopes7 commented on receiver/netflowreceiver/producer.go:
```diff
@@ -0,0 +1,94 @@
+	}
+
+	// Send the logs to the collector, it is difficult to pass the context here
+	err = o.logConsumer.ConsumeLogs(context.TODO(), log)
```
Implemented [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36865/commits/169373f9f6f3584468a4d1987346a71ba6933605)

dlopes7 commented on receiver/netflowreceiver/receiver.go:
```diff
@@ -5,25 +5,165 @@ package netflowreceiver // import "github.com/open-telemetry/opentelemetry-colle
 	"go.uber.org/zap"
 )
 
+type dropHandler struct {
```
Implemented [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36865/commits/5cf7625a031808b396fec5f1ece9ac700e9c920a)

dlopes7 commented on receiver/netflowreceiver/receiver.go:
```diff
@@ -5,25 +5,165 @@ package netflowreceiver // import "github.com/open-telemetry/opentelemetry-colle
+		return err
+	}
+
+	nr.logger.Info("Start listening for NetFlow on UDP", zap.Any("config", nr.config))
```
Implemented [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36865/commits/54d8b60e83623ea11b5a48a17573827328e8b8b3)

dlopes7 committed:
netflow - sflow and netflow schema documentation

* Log empty sflow packets, which can happen if a packet only contains counter samples, which are not yet supported

dlopes7 committed:
netflow - remove 'flow' scheme

The flow scheme was strange as it was just netflow + sflow in a single pipe which checked the protocol before forwarding tp the actual pipe

We remove that option for clarity, so users can only chose netflow or sflow as the scheme

dlopes7 committed:
netflow - add  sflow to sample yaml config

dlopes7 committed:
netflow - fix default queue size

dlopes7 committed:
Merge remote-tracking branch 'upstream/main' into netflow-receiver-implementation

dlopes7 committed:
netflow - go mod tidy

dlopes7 committed:
netflow - format

dlopes7 commented:
@evan-bradley would you mind taking a look at the changes after your review please?

Eventually we want to be more explicit on how we parse the network packets instead of relying on the goflow2 proto producer.

This way we can parse things like counters from sflow (which are probably better as metrics instead of logs)

For now I've documented the supported schemas and what they can do, thanks for the in depth review!

dlopes7 committed:
netflow - lint and typing

dlopes7 committed:
Merge remote-tracking branch 'upstream/main' into netflow-receiver-implementation

dlopes7 committed:
Merge remote-tracking branch 'upstream/main' into netflow-receiver-implementation

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

evan-bradley commented on receiver/netflowreceiver/receiver.go:
```diff
@@ -5,25 +5,164 @@ package netflowreceiver // import "github.com/open-telemetry/opentelemetry-colle
 }
 
 func (nr *netflowReceiver) Start(_ context.Context, _ component.Host) error {
+	nr.logger.Info("NetFlow receiver is starting...")
+
+	// The function that will decode packets
```
```suggestion
	// The function that will decode packets
```
The Collector more or less already logs this, so I think we can remove this log.

evan-bradley commented on receiver/netflowreceiver/receiver.go:
```diff
@@ -5,25 +5,164 @@ package netflowreceiver // import "github.com/open-telemetry/opentelemetry-colle
+		return err
+	}
+
+	nr.logger.Info("Start listening over UDP", zap.String("scheme", nr.config.Scheme), zap.Int("port", nr.config.Port))
```
```suggestion
	nr.logger.Info("Starting UDP listener", zap.String("scheme", nr.config.Scheme), zap.Int("port", nr.config.Port))
```
Minor wording suggestion

evan-bradley commented on receiver/netflowreceiver/receiver.go:
```diff
@@ -5,25 +5,164 @@ package netflowreceiver // import "github.com/open-telemetry/opentelemetry-colle
 
-func (nr *netflowReceiver) Shutdown(_ context.Context) error {
+func (nr *netflowReceiver) Shutdown(context.Context) error {
+	nr.logger.Info("NetFlow receiver is shutting down...")
```
```suggestion
```
Same as with startup, I think we can avoid logging here since the Collector prints logs saying it is shutting down.

evan-bradley reviewed (commented):
This is looking pretty good now, thanks for incorporating all the feedback.

jotak commented on receiver/netflowreceiver/README.md:
```diff
@@ -61,42 +66,52 @@ You would then configure your network devices to send netflow, sflow, or ipfix d
+* **network.type**: Str(ipv4)
+* **network.io.bytes**: Int(853)
+* **network.io.packets**: Int(83)
+* **network.flow.type**: Str(netflow_v5)
+* **network.flow.sequence_num**: Int(191)
+* **network.flow.time_received**: Int(1736309689918929427)
+* **network.flow.start**: Int(1736309689830846400)
+* **network.flow.end**: Int(1736309689871846400)
+* **network.flow.sampling_rate**: Int(0)
+* **network.flow.sampler_address**: Str(172.28.176.1)
```
is there any plan/possibility to promote these as new semconv sooner or later?

dlopes7 replied:
I am not familiar with the process of suggesting new entries to the semconv, but I think it would be a good idea as these are pretty universal

dlopes7 committed:
netflow - better wording on log

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
netflow - rename fields

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
netflow - remove unecessary log line

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
netflow - better reference in comment

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
netflow - remove unecessary log line

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
netflow - move flow keys to 'flow.' namespace

dlopes7 committed:
netflow - update readme keys

dlopes7 commented on receiver/netflowreceiver/parser.go:
```diff
@@ -0,0 +1,252 @@
+	// There is no semconv as of today for these
+	// Network IO attributes
+	r.Attributes().PutInt("network.io.bytes", int64(pm.Bytes))
+	r.Attributes().PutInt("network.io.packets", int64(pm.Packets))
```
Implemente [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36865/commits/64aed8cb4c1a7c7c8100bec2ce7b91ad676a2c47)

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

dlopes7 committed:
netflow - adjust validSchemes slice size

strawgate commented on receiver/netflowreceiver/config_test.go:
```diff
@@ -36,7 +36,27 @@ func TestLoadConfig(t *testing.T) {
+				Port:      2055,
+				Sockets:   1,
+				Workers:   1,
+				QueueSize: 1000,
```
is this supposed to be 0 in the zero_queue test case?

dlopes7 committed:
netflow - clarify minimum queue size

dlopes7 committed:
netflow - simplify panic handling

dlopes7 commented on receiver/netflowreceiver/receiver.go:
```diff
@@ -5,25 +5,164 @@ package netflowreceiver // import "github.com/open-telemetry/opentelemetry-colle
+	decodeFunc = p.DecodeFlow
+
+	// We wrap panics while decoding the message to handle them later
+	decodeFunc = debug.PanicDecoderWrapper(decodeFunc)
```
@evan-bradley I've removed that panic wrapper and used a simple recover in the produce function [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36865/commits/44e134e7c9e7e9915034835f0a36e04bad184751)

dlopes7 committed:
netflow - formatting and remove a comment

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

evan-bradley commented on receiver/netflowreceiver/producer.go:
```diff
@@ -0,0 +1,84 @@
+
+// Produce converts the message into a list log records and sends them to log consumer
+func (o *OtelLogsProducerWrapper) Produce(msg any, args *producer.ProduceArgs) ([]producer.ProducerMessage, error) {
+	if pErr := recover(); pErr != nil {
+		errMessage, _ := pErr.(string)
+		o.logger.Error("unexpected error processing the message", zap.String("error", errMessage))
+		return nil, ErrorProducerProcessing
+	}
```
```suggestion
	defer func() {
		if pErr := recover(); pErr != nil {
			errMessage, _ := pErr.(string)
			o.logger.Error("unexpected error processing the message", zap.String("error", errMessage))
			return nil, ErrorProducerProcessing
		}
	}
```
The panic recovery has to happen in a deferred function execution to work. Also, is there a way we can test this?

dlopes7 replied:
Sorry I missed this obvious mistake, writing tests for it, I think we can create a producer that panics, and use it in place of the proto producer on the test

The only panic here would come from the proto producer implementation

dlopes7 replied:
Added it [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36865/commits/af1e72bd75a26de470fba6bfd737a4a5625c2f29)

dlopes7 committed:
Update receiver/netflowreceiver/producer.go

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
Update receiver/netflowreceiver/producer.go

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

dlopes7 committed:
netflow - fix panic handling and test for it

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

dlopes7 committed:
Merge branch 'netflow-receiver-implementation' of github.com:dlopes7/opentelemetry-collector-contrib into netflow-receiver-implementation

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

evan-bradley commented on receiver/netflowreceiver/parser_test.go:
```diff
@@ -0,0 +1,118 @@
+	expectedAttributes.PutInt("flow.sampling_rate", 1)
+	expectedAttributes.PutStr("flow.sampler_address", "192.168.1.100")
+
+	assert.Equal(t, expectedAttributes, record.Attributes())
```
We should also check the log record timestamps in this test and the one below.

dlopes7 replied:
Great idea, and they were actually inverted (timestamp and observed timestamp), which came up during writing this test

dlopes7 committed:
netflow - add tests for parsed timestamps

dlopes7 committed:
Merge remote-tracking branch 'upstream/main' into netflow-receiver-implementation

dlopes7 committed:
netflow - go mod tidy

dlopes7 committed:
netflow - update semconv

dlopes7 committed:
Merge branch 'main' into netflow-receiver-implementation

evan-bradley committed:
Merge remote-tracking branch 'upstream/main' into netflow-receiver-implementation

dlopes7 committed:
netflow - update semconv

evan-bradley committed:
go mod tidy
